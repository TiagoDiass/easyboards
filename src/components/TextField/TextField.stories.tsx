import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Email as EmailIcon } from 'styled-icons/material-outlined';
import TextField from './TextField';

export default {
  title: 'TextField',
  component: TextField,
  args: {
    label: 'Email',
    labelFor: 'email',
    initialValue: '',
    placeholder: 'jhon.doe@gmail.com'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: 'symbol' }
  }
} as ComponentMeta<typeof TextField>;

export const Basic: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
);

export const WithIcon: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} autoComplete='off' />
  </div>
);

WithIcon.args = {
  icon: <EmailIcon />
};

export const Disabled: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
);

Disabled.args = {
  disabled: true,
  value: 'tiago@email.com'
};

export const WithError: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
);

WithError.args = {
  error: 'Invalid email',
  value: 'invalid-email@test.com',
  icon: <EmailIcon />
};
