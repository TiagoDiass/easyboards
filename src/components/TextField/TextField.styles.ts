import styled, { css, DefaultTheme } from 'styled-components';
import { TextFieldProps } from './TextField';

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label}, ${Input}, ${Icon} {
      color: ${theme.colors.primary.accent3};
      cursor: not-allowed;

      &::placeholder {
        color: currentColor;
      }
    }
  `,

  withError: (theme: DefaultTheme) => css`
    ${Label}, ${Icon} {
      color: ${theme.colors.error.default};
    }

    ${InputWrapper} {
      border: 1px solid ${theme.colors.error.default};

      &:focus-within {
        border: 2px solid ${theme.colors.error.darker};
        box-shadow: none;
      }
    }
  `
};

type WrapperProps = {
  disabled: TextFieldProps['disabled'];
  hasError: boolean;
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, hasError }) => css`
    display: flex;
    flex-direction: column;
    row-gap: ${theme.spacings.xxsmall};

    ${hasError && wrapperModifiers.withError(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.secondary.accent2};
    cursor: pointer;
  `}
`;

const inputWrapperModifiers = {
  iconOnRight: () => css`
    flex-direction: row-reverse;
  `
};

type InputWrapperProps = Pick<TextFieldProps, 'iconPosition'>;

export const InputWrapper = styled.div<InputWrapperProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.primary.accent1};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xxsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.primary.accent2};

    transition: box-shadow ${theme.transitions.default};

    &:focus-within {
      border: 1px solid ${theme.colors.primary.accent3};
    }

    ${iconPosition === 'right' && inputWrapperModifiers.iconOnRight()}
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.secondary.accent2};
    /* font-family: ${theme.font}; */
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    ::placeholder {
      color: red;
      color: ${theme.colors.primary.accent3};
    }

    /* Removes the autofill styles from browser */
    &:-webkit-autofill,
    &:-webkit-autofill:hover,
    &:-webkit-autofill:focus,
    &:-webkit-autofill:active {
      -webkit-box-shadow: 0 0 0 30px ${theme.colors.primary.accent1} inset !important;
    }
  `}
`;

export const Icon = styled.div`
  ${({ theme }) => css`
    display: flex;
    width: 2.2rem;
    color: ${theme.colors.primary.accent3};

    & > svg {
      width: 100%;
    }
  `}
`;

export const Error = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.error.default};
    font-size: ${theme.font.sizes.xsmall};
  `}
`;
