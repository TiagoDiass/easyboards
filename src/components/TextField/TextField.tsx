import { ChangeEvent, InputHTMLAttributes, useState } from 'react';
import * as S from './TextField.styles';

export type TextFieldProps = {
  onInput?: (value: string) => void;
  label?: string;
  labelFor?: string;
  initialValue?: string;
  disabled?: boolean;

  /**
   * Error message to let the user know that there is an error with the provided input.
   *
   * PS: you should provide the `labelFor` prop when using this error prop, to make the component more accessible
   */
  error?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
} & InputHTMLAttributes<HTMLInputElement>;

/**
 * TextField component that is used for user inputs
 */
export default function TextField({
  label,
  labelFor,
  initialValue = '',
  onInput,
  disabled = false,
  error,
  icon,
  iconPosition = 'left',
  ...props
}: TextFieldProps) {
  const [value, setValue] = useState(initialValue);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;

    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  const hasError = !!error;
  const errorElementId = labelFor && `${labelFor}-error`;

  return (
    <S.Wrapper disabled={disabled} hasError={hasError}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}

      <S.InputWrapper iconPosition={iconPosition}>
        {!!icon && <S.Icon>{icon}</S.Icon>}
        <S.Input
          type='text'
          id={labelFor}
          value={value}
          onChange={handleChange}
          disabled={disabled}
          aria-disabled={disabled}
          aria-invalid={hasError ? 'true' : 'false'}
          aria-errormessage={errorElementId}
          {...props}
        />
      </S.InputWrapper>

      {hasError && <S.Error id={errorElementId}>{error}</S.Error>}
    </S.Wrapper>
  );
}
