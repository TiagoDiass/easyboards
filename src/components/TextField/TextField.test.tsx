import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/test-utils';
import TextField from './TextField';
import { Email as EmailIcon } from 'styled-icons/material-outlined';

describe('<TextField />', () => {
  it('should render correctly with Label', () => {
    renderWithTheme(<TextField label='Label' labelFor='Field' />);

    expect(screen.getByLabelText('Label')).toBeInTheDocument();
  });

  it('should render correctly without Label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
  });

  it('should render correctly with placeholder', () => {
    renderWithTheme(<TextField placeholder='Type your email in here' />);

    expect(screen.getByPlaceholderText('Type your email in here')).toBeInTheDocument();
  });

  it('should render correctly with an icon', () => {
    renderWithTheme(
      <TextField label='Label' labelFor='Field' icon={<EmailIcon data-testid='icon' />} />
    );

    expect(screen.getByLabelText('Label')).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render correctly with an icon on the right side', () => {
    renderWithTheme(
      <TextField label='Label' labelFor='Field' icon={<EmailIcon />} iconPosition='right' />
    );

    expect(screen.getByLabelText('Label').parentElement).toHaveStyle({
      display: 'flex',
      flexDirection: 'row-reverse'
    });
  });

  it('should disable the input when disabled prop is specified', () => {
    renderWithTheme(
      <TextField
        disabled
        label='Label'
        labelFor='Field'
        icon={<EmailIcon />}
        iconPosition='right'
      />
    );

    const inputElement = screen.getByRole('textbox');

    expect(inputElement).toBeDisabled();
    expect(inputElement).toHaveAttribute('aria-disabled', 'true');
  });

  it('should render correctly with an accessible error', () => {
    const inputId = 'email';
    const errorMessage = 'Please fill out this field';

    const { container } = renderWithTheme(
      <TextField label='Label' labelFor={inputId} error={errorMessage} />
    );

    const errorElement = screen.getByText(errorMessage);
    const inputElement = screen.getByRole('textbox', { name: 'Label' });

    expect(inputElement).toHaveAttribute('aria-invalid', 'true');
    expect(inputElement).toHaveAttribute('aria-errormessage', `${inputId}-error`);

    expect(errorElement).toBeInTheDocument();
    expect(errorElement).toHaveAttribute('id', `${inputId}-error`);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call props.onInput correctly when input value changes', async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField onInput={onInput} label='TextField' labelFor='TextField' id='TextField' />
    );

    const input = screen.getByRole('textbox');
    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });
    expect(onInput).toHaveBeenLastCalledWith(text);
  });

  it('should be accessible with the tab key', () => {
    renderWithTheme(<TextField label='TextField' labelFor='TextField' id='TextField' />);

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });
});
