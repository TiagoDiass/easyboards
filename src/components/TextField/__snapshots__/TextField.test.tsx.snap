// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render correctly with an accessible error 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 0.8rem;
}

.c0 .c2 {
  border: 1px solid #FF1A1A;
}

.c0 .c2:focus-within {
  border: 2px solid #E00;
  box-shadow: none;
}

.c1 {
  font-size: 1.4rem;
  color: #333;
  cursor: pointer;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #FAFAFA;
  border-radius: 0.2rem;
  padding: 0 0.8rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.c3:focus-within {
  border: 1px solid #999;
}

.c4 {
  color: #333;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c4::-webkit-input-placeholder {
  color: red;
  color: #999;
}

.c4::-moz-placeholder {
  color: red;
  color: #999;
}

.c4:-ms-input-placeholder {
  color: red;
  color: #999;
}

.c4::placeholder {
  color: red;
  color: #999;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus,
.c4:-webkit-autofill:active {
  -webkit-box-shadow: 0 0 0 30px #FAFAFA inset !important;
}

.c5 {
  color: #FF1A1A;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="email"
  >
    Label
  </label>
  <div
    class="c2 c3"
  >
    <input
      aria-disabled="false"
      aria-errormessage="email-error"
      aria-invalid="true"
      class="c4"
      id="email"
      type="text"
      value=""
    />
  </div>
  <span
    class="c5"
    id="email-error"
  >
    Please fill out this field
  </span>
</div>
`;
